src/geff/networkx/io.py

```diff
@@ -0,0 +1,159 @@
+
+
+def get_edge_attrs(graph: nx.Graph) -> list[str]:
+    """Get the attribute keys present on any edge in the networkx graph. Does not imply
+    that the attributes are present on all edges.
+
+    Args:
+        graph (nx.Graph): a networkx graph
+
+    Returns:
+        list[str]: A list of all unique edge attribute keys
+    """
+    return list({k for e in graph.edges for k in graph.edges[e]})
+
+
+def write(
+    graph: nx.Graph,
+    position_attr: str,
+    path: str | Path,
+    axis_names: Optional[list[str]] = None,
+    axis_units: Optional[list[str]] = None,
+):
+    """Write a networkx graph to the geff file format
+
+    Args:
+        graph (nx.Graph): a networkx graph where every node has a position attribute
+        position_attr (str): the name of the position attribute present on every node
+        path (str | Path): the path to the output zarr. Opens in append mode,
+            so will only overwrite geff-controlled groups.
+        axis_names (Optional[list[str]], optional): The names of the spatial dims
+            represented in position attribute. Defaults to None.
+        axis_units (Optional[list[str]], optional): The units of the spatial dims
+            represented in position attribute. Defaults to None.
+    """
+    if nx.is_empty(graph):
+        warnings.warn(f"Graph is empty - not writing anything to {path}", stacklevel=2)
+        return
+    # open/create zarr container
+    group = zarr.open(path, "a")
+
+    # write meta-data
+    group.attrs["geff_spec"] = geff.__version__
+    group.attrs["position_attr"] = position_attr
+    group.attrs["directed"] = isinstance(graph, nx.DiGraph)
+    group.attrs["roi"] = get_roi(graph, position_attr=position_attr)
+    if axis_names:
+        graph.attrs["axis_names"] = axis_names
+    if axis_units:
+        graph.attrs["axis_units"] = axis_units
+
+    # get node and edge IDs
+    nodes_list = list(graph.nodes())
+    nodes_arr = np.array(nodes_list)
+    edges_list = list(graph.edges())
+    edges_arr = np.array(edges_list)
+
+    # write nodes
+    group["nodes/ids"] = nodes_arr
+
+    # write node attributes
+    for name in get_node_attrs(graph):
+        # TODO: handle missing values
+        group[f"nodes/attrs/{name}"] = np.array([graph.nodes[node][name] for node in nodes_list])
+
+    # write edges
+    group["edges/ids"] = edges_arr
+
+    # write edge attributes
+    for name in get_edge_attrs(graph):
+        group[f"edges/attrs/{name}"] = np.array([graph.edges[edge][name] for edge in edges_list])
+
+
+def read(path: Path) -> nx.Graph:
+    """Read a geff file into a networkx graph.
+
+    Args:
+        path (Path): The path to the root of the geff zarr, where the .attrs contains
+            the geff  metadata
+    Returns:
+        nx.Graph: The graph that was stored in the geff file format
+    """
+    # open zarr container
+    path = Path(path)
+    assert path.exists(), f"Cannot read graph from {path} because it does not exist"
```

Comment:
We could use the validation function here to check that the file is ok. Probably under a flag so that it can be disabled.